module/irrecoverable/irrecoverable.go

```diff
@@ -6,30 +6,43 @@ import (
 	"log"
 	"os"
 	"runtime"
+
+	"go.uber.org/atomic"
 )
 
-// Signaler sends the error out
+// Signaler sends the error out.
 type Signaler struct {
-	errors chan<- error
+	errChan   chan error
+	errThrown *atomic.Bool
+}
+
+func NewSignaler() *Signaler {
+	return &Signaler{
+		errChan:   make(chan error, 1),
+		errThrown: atomic.NewBool(false),
+	}
 }
 
-func NewSignaler(errors chan<- error) *Signaler {
-	return &Signaler{errors}
+// Error returns the Signaler's error channel.
+func (s *Signaler) Error() <-chan error {
+	return s.errChan
 }
 
 // Throw is a narrow drop-in replacement for panic, log.Fatal, log.Panic, etc
-// anywhere there's something connected to the error channel
-func (e *Signaler) Throw(err error) {
-	defer func() {
-		// If the error channel was already closed by a concurrent call to Throw, the call
-		// to close below will panic. We simply log the unhandled irrecoverable for now.
-		if r := recover(); r != nil {
-			log.New(os.Stderr, "", log.LstdFlags).Println(fmt.Errorf("unhandled irrecoverable: %w", err))
-		}
-		runtime.Goexit()
-	}()
-	e.errors <- err
-	close(e.errors)
+// anywhere there's something connected to the error channel. It only sends
+// the first error it is called with to the error channel, and logs subsequent
+// errors as unhandled.
+func (s *Signaler) Throw(err error) {
+	defer runtime.Goexit()
+
+	if s.errThrown.CAS(false, true) {
+		s.errChan <- err
+		close(s.errChan)
+	} else {
+		// TODO: we simply log the unhandled irrecoverable for now, but we should probably
+		// find a way to actually handle them.
+		log.New(os.Stderr, "", log.LstdFlags).Println(fmt.Errorf("unhandled irrecoverable: %w", err))
+	}
```

Comment:
I'm not against this, and in the very first version of the error handling PR I had implemented things under this assumption. The reason I made this chance is actually because I realized there was an inconsistency with how we ended up implementing irrecoverable API and RunComponent.

If we choose to go with this approach, it means we should remove [these lines](https://github.com/onflow/flow-go/blob/smnzhu%2Fmiddleware-component/module/component/component.go#L79-L83) for consistency.

The way it was previously implemented, RunComponent shuts down the component upon the first irrecoverable. If we decide we want to keep the channel open, then RunComponent should not automatically shut down the component, and should delegate that responsibility to the error handler. If the error handler does not cancel the component, we should keep listening for irrecoverables on the channel until it does.

Reply:
But the thing is, I don't think this approach is compatible with having the error handler being defined as a function which takes in a single error, rather than a function which takes in an error channel and a restart function, which is what I originally had implemented.

Here are some edge cases:
* If the handler restarts a component, should we then ignore the remaining irrecoverables in the channel? I would say probably not, because the handler may still want to do things with those irrecoverables, such as logging them somewhere. However, if the handler is defined as a single function, then how does it know whether an irrecoverable is from the current component instance or is one leftover from a previous run? 
* If the handler cancels the context, signaling us not to restart the component, does this also mean we should drop the remaining irrecoverables? I still think the answer is no here.

I think we can either: 
* Declare that a signaler will only propagate the first error it receives. This would be compatible with the current implementation of RunComponent and ComponentManager. I think that this is a valid way to define the API, since an irrecoverable error should be an error that is severe enough that it warrants crashing the entire component. After all, the current behaviour across the codebase calls `log.Fatal`, which crashes the entire program.
* Let the signaler continue propagating errors to the handler indefinitely (until the handler decides it stops wanting to receive them or until there are no more to receive because the component has shut down). This would mean that the error handler API should be modified to take in an error channel and a restart function (which is idempotent and only triggers restart once), so that it is clear that all the errors within the channel are within the scope of a single instance of the component. This is also a valid way to define the API.

Reply:
OK, seen like this, I think I'm fine with the "let it crash" philosophy. Errors are either irrecoverable or they are not. If they are, the most one would want to do with the ones occurring after the first one is log them. 

As a consequence, the current setup is probably fine, and much simpler than the alternative. But this "kill on first irrecoverable" particularity should not surprise users, and probably deserves a mention in the FLIP